### ISSUE TITLE
On using dash::Matrix

### ISSUE THREAD

**Issue by devreal** [dash-project/dash#412]
I poked around with `dash::Matrix` (or `dash::NArray` to be precise) today and want to report some of my findings. Consider the following code:

```
#include <libdash.h>
#include <vector>

int main(int argc, char **argv)
{
  constexpr int elem_per_unit = 1000;
  dash::init(&argc, &argv);

  std::vector<double> vec(100);
  std::fill(vec.begin(), vec.end(), dash::myid());

  dash::NArray<double, 2> matrix;

  dash::SizeSpec<2> sizespec(dash::size(), elem_per_unit);
  dash::DistributionSpec<2> distspec(dash::BLOCKED, dash::NONE);
  dash::TeamSpec<2> teamspec(dash::size(), 1);

  matrix.allocate(sizespec, distspec, teamspec);
  if (dash::myid() == 0) {
    // ostream<< overload for iterators is broken
    // std::cout << matrix.row(1).begin() << std::endl;
    dash::copy(&vec[0], &vec[0] + vec.size(), matrix.row(1).begin() + 100);
  }

  matrix.barrier();
  dash::finalize();

  return 0;
}
```

My issues:
1. *Delayed allocation*: why do I have to create a `SizeSpec`, `DistSpec`, and `TeamSpec` if for direct allocation through the c'tor passing the size in each dimension is sufficient? I also noticed that the types (`SizeSpec_t` etc) used in the interface of `allocate()` are private, yikes...
2. The goal of the exercise was to copy data to a *random position* in a matrix. My initial attempt was to pass something like `&matrix[1][100]` to `dash::copy`, because, hey, I can also do `&vec[0]`, i.e., take a pointer to a reference. So why not provide `operator&` on `dash::GlobRef` to return a `dash::GlobPtr`, that can be passed to `dash::copy`? The current solution is not exactly intuitive. 
3. The `ostream<<` on iterators is broken, leading to a weird error: 
```
In file included from /home/joseph/opt/dash-0.3.0/include/dash/iterator/GlobIter.h:742:0,
                 from /home/joseph/opt/dash-0.3.0/include/dash/Iterator.h:6,
                 from /home/joseph/opt/dash-0.3.0/include/libdash.h:26,
                 from dash_local_to_matrix.cc:1:
/home/joseph/opt/dash-0.3.0/include/dash/iterator/GlobViewIter.h: In instantiation of ‘std::ostream& dash::operator<<(std::ostream&, const dash::GlobViewIter<ElementType, Pattern, GlobStaticMem, Pointer, Reference>&) [with ElementType = double; Pattern = dash::BlockPattern<2, (dash::MemArrange)1u, long int>; GlobStaticMem = dash::GlobStaticMem<double, dash::allocator::SymmetricAllocator<double> >; Pointer = dash::GlobPtr<double, dash::GlobStaticMem<double, dash::allocator::SymmetricAllocator<double> > >; Reference = dash::GlobRef<double>; std::ostream = std::basic_ostream<char>]’:
dash_local_to_matrix.cc:21:38:   required from here
/home/joseph/opt/dash-0.3.0/include/dash/iterator/GlobViewIter.h:1094:51: error: no matching function for call to ‘dash::GlobPtr<double, dash::GlobStaticMem<double, dash::allocator::SymmetricAllocator<double> > >::GlobPtr(const dash::GlobViewIter<double, dash::BlockPattern<2, (dash::MemArrange)1u, long int>, dash::GlobStaticMem<double, dash::allocator::SymmetricAllocator<double> >, dash::GlobPtr<double, dash::GlobStaticMem<double, dash::allocator::SymmetricAllocator<double> > >, dash::GlobRef<double> >&)’
   dash::GlobPtr<ElementType, GlobStaticMem> ptr(it);
```
4. DASH and DART are dicks if not initialized properly (which of course has it's reasons...)

Just wanted to document these and hopefully draw something useful out of my experience.

**Comment by fuchsto** [comment]
Local-to-global and global-to-global copying is only "officially" supported with PR #410 (also see https://github.com/dash-project/dash/issues/346), I will add your use case to a test suite there.

**Comment by devreal** [comment]
Points 1 and 3 have been addressed in #417. Number 4 is not likely to change as checking for initialization adds some overhead. 

Regarding point 3: I tried to address this as well but realized that we need to carry a pointer to the underlying global memory instance in each `GlobRef`, which would then be passed to the constructor of `GlobPtr` in `operator&()`. While I like the convenience of `&matrix[1][100]`, I'm unsure whether the increase in size of `GlobRef` justifies this (probably rather rare) use-case. IMO, the current solution(s) are not ideal, either.
